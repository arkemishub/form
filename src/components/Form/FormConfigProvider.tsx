import React, { createContext, useContext } from "react";
import {
  FormConfigComponents,
  FormConfigProviderProps,
  FormConfigValue,
} from "../../types";

const FormConfigContext = createContext<FormConfigValue>({
  components: {} as FormConfigComponents,
});

export function FormConfigProvider({
  components,
  children,
}: FormConfigProviderProps) {
  return (
    <FormConfigContext.Provider
      value={{ components: components ?? ({} as FormConfigComponents) }}
    >
      {children}
    </FormConfigContext.Provider>
  );
}

export const useFormConfig = () => useContext(FormConfigContext);
