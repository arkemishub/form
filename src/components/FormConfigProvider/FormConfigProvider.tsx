import React, { createContext, useContext } from "react";
import type { FormComponents } from "../../types";
import type {
  FormConfigProviderProps,
  FormConfigValue,
} from "./FormConfigProvider.types";

const FormConfigContext = createContext<FormConfigValue>({
  components: {} as FormComponents,
});

function FormConfigProvider({ components, children }: FormConfigProviderProps) {
  return (
    <FormConfigContext.Provider
      value={{ components: components ?? ({} as FormComponents) }}
    >
      {children}
    </FormConfigContext.Provider>
  );
}

const useFormConfig = () => useContext(FormConfigContext) ?? {};

export default FormConfigProvider;
export { useFormConfig };
