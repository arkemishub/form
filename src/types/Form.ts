/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import {
  CSSProperties,
  PropsWithChildren,
  ReactElement,
  ReactNode,
} from "react";
import { RenderProps } from "./RenderProps";
import { FieldValues, UseFormReturn } from "react-hook-form";
import { Field, FieldType } from "./Field";

type FormConfigComponents = Partial<
  Record<FieldType, (props: RenderProps) => ReactElement>
>;

type FormConfigProviderProps = PropsWithChildren<{}> & {
  components?: FormConfigComponents;
};

type FormConfigValue = {
  components: FormConfigComponents;
};

type FormValue = {
  components: FormConfigComponents;
  fields: Field[];
  onChange?(data: FieldValues): void;
};

type FormProps = {
  id?: string;
  children: ReactNode;
  onChange?(data: FieldValues): void;
  onSubmit?(data: FieldValues): void;
  components?: FormConfigComponents;
  fields: Field[];
  style?: CSSProperties;
  methods?: UseFormReturn;
};

export {
  FormConfigComponents,
  FormConfigProviderProps,
  FormConfigValue,
  FormValue,
  FormProps,
};
