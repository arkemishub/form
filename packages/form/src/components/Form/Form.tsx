/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { FormProvider } from "react-hook-form";
import {
  Children,
  cloneElement,
  isValidElement,
  ReactElement,
  ReactNode,
  useCallback,
  useMemo,
} from "react";
import { MarkRequired } from "../../types/utils";
import FormField from "../FormField/FormField";
import { useFormConfig } from "../FormConfigProvider/FormConfigProvider";
import { FormProps, OuterFormProps } from "./Form.types";
import { useForm } from "../../hooks";

function FormComponent({
  id,
  onSubmit,
  className,
  style,
  methods: { handleSubmit },
  onChange,
  fields,
  children,
  ...props
}: MarkRequired<FormProps, "methods">) {
  const config = useFormConfig();
  const components = useMemo(
    () => ({ ...config.components, ...props.components }),
    [config.components, props.components]
  );

  const buildChildren = useCallback(
    (children: ReactNode): ReactNode => {
      return Children.map(children, (child) => {
        if (isValidElement(child)) {
          const reactChild = child as ReactElement;

          if (reactChild?.type === FormField) {
            return cloneElement(reactChild, {
              components,
              fields,
              onChange,
            });
          }

          if (reactChild.props?.children) {
            return cloneElement(reactChild, {
              ...reactChild.props,
              children: buildChildren(reactChild.props.children),
            });
          }
        }

        return child;
      });
    },
    [fields, components, onChange]
  );

  return (
    <form
      id={id}
      onSubmit={onSubmit ? handleSubmit(onSubmit) : undefined}
      className={className}
      style={style}
    >
      {buildChildren(children)}
    </form>
  );
}

function Form({ getFieldDefaultValue, ...props }: OuterFormProps) {
  const { methods: defaultMethods } = useForm(
    props?.fields
      ? {
          fields: props.fields,
          getFieldDefaultValue,
        }
      : undefined
  );
  const methods = useMemo(
    () => props.methods ?? defaultMethods,
    [props.methods, defaultMethods]
  );

  return (
    <FormProvider {...methods}>
      <FormComponent {...props} methods={methods} />
    </FormProvider>
  );
}

Form.Field = FormField;
export default Form;
