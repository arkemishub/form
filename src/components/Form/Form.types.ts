/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { FieldValues, SubmitHandler, UseFormReturn } from "react-hook-form";
import { CSSProperties, PropsWithChildren } from "react";
import { Field, FormComponents } from "../../types";

type FormProps<TFieldValues extends FieldValues = FieldValues> =
  PropsWithChildren<{
    id?: string;
    methods?: UseFormReturn<TFieldValues>;
    onSubmit?: SubmitHandler<TFieldValues>;
    style?: CSSProperties;
    className?: string;
    components?: FormComponents;
    fields?: Field[];
    onChange?: (event: { target: any; type?: any }) => void;
  }>;

export type { FormProps };
