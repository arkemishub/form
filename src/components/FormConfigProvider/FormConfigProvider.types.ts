import { PropsWithChildren } from "react";
import { FormComponents } from "../../types";

type FormConfigProviderProps = PropsWithChildren<{}> & {
  components?: FormComponents;
};

type FormConfigValue = {
  components: FormComponents;
};

export type { FormConfigValue, FormConfigProviderProps };
