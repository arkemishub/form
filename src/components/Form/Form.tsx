/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { FormProvider, useForm } from "react-hook-form";
import {
  Children,
  cloneElement,
  isValidElement,
  ReactElement,
  useMemo,
} from "react";
import { MarkRequired } from "../../types/utils";
import FormField from "../FormField/FormField";
import { useFormConfig } from "../FormConfigProvider/FormConfigProvider";
import { FormProps } from "./Form.types";

function FormComponent({
  id,
  onSubmit,
  className,
  style,
  methods: { handleSubmit },
  onChange,
  fields,
  ...props
}: MarkRequired<FormProps, "methods">) {
  const config = useFormConfig();
  const components = useMemo(
    () => ({ ...config.components, ...props.components }),
    [config.components, props.components]
  );

  const children = useMemo(() => {
    return Children.map(props.children, (child) => {
      if (isValidElement(child)) {
        if (child?.type === FormField) {
          return cloneElement(child as ReactElement, {
            components,
            fields,
            onChange,
          });
        }

        return child;
      }
    });
  }, [props.children]);

  return (
    <form
      id={id}
      onSubmit={onSubmit ? handleSubmit(onSubmit) : undefined}
      className={className}
      style={style}
    >
      {children}
    </form>
  );
}

function Form(props: FormProps) {
  const defaultMethods = useForm();
  const methods = useMemo(
    () => props.methods ?? defaultMethods,
    [props.methods, defaultMethods]
  );

  return (
    <FormProvider {...methods}>
      <FormComponent {...props} methods={methods} />
    </FormProvider>
  );
}

Form.Field = FormField;
export default Form;
