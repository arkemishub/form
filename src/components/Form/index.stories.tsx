/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ReactNode, useState } from "react";
import mockData from "../../__mocks__/mockData";
import { Field } from "../../types";
import Form from "./Form";
import { FormConfigProvider } from "../FormConfigProvider";
import { useForm } from "react-hook-form";

export default {
  title: "Form",
  component: Form,
};

const GeneralFormProvider = ({ children }: { children: ReactNode }) => {
  return (
    <FormConfigProvider
      components={{
        boolean: ({ field }) => (
          <div style={{ display: "flex" }}>
            <input
              {...field}
              type="checkbox"
              checked={field.value}
              onChange={(e) => field.onChange(e.target.checked)}
            />
            <div>{field.label}</div>
          </div>
        ),
        string: ({ field }) => (
          <>
            <input
              {...field}
              placeholder={field.label}
              onChange={(e) => field.onChange(e.target.value)}
            />
          </>
        ),
        // default: () => <div>Component not found</div>,
      }}
    >
      {children}
    </FormConfigProvider>
  );
};

export const Default = () => {
  const methods = useForm();
  const [fields] = useState<Field[]>(mockData);
  const [submitData, setSubmitData] = useState({});

  return (
    <>
      <GeneralFormProvider>
        <Form
          methods={methods}
          fields={fields}
          onSubmit={(values) => setSubmitData(values)}
        >
          <div>
            <div
              style={{
                display: "grid",
                gridTemplateColumns: "auto auto auto auto",
                gridGap: "8px 20px",
              }}
            >
              {fields.map((field: { id: string }) => (
                <Form.Field key={`field-${field.id}`} id={field.id} />
              ))}
            </div>
            <button style={{ marginTop: 20 }}>Submit</button>
          </div>
        </Form>
      </GeneralFormProvider>
      <br />
      {JSON.stringify(submitData)}
    </>
  );
};

export const WithFormProvider = () => {
  const methods = useForm();

  const [fields] = useState<Field[]>(mockData);
  const [submitData, setSubmitData] = useState({});

  return (
    <>
      <GeneralFormProvider>
        <Form fields={fields} onSubmit={(values) => setSubmitData(values)}>
          <div>
            <div
              style={{
                display: "grid",
                gridTemplateColumns: "auto auto auto auto",
                gridGap: "8px 20px",
              }}
            >
              {fields.map((field: { id: string }) => (
                <Form.Field key={`field-${field.id}`} id={field.id} />
              ))}
            </div>
            <button style={{ marginTop: 20 }}>Submit</button>
          </div>

          <div style={{ marginTop: 20, display: "flex", gap: 4 }}>
            <button type="button" onClick={() => methods.reset({})}>
              Reset
            </button>
            <button type="button" onClick={() => methods.setValue("name", "")}>
              Clean name field
            </button>
            <button type="button" onClick={() => methods.register("name")}>
              register name field
            </button>
            <button type="button" onClick={() => methods.unregister("name")}>
              Unregister name field
            </button>
          </div>
        </Form>
      </GeneralFormProvider>
      <br />
      {JSON.stringify(submitData)}
    </>
  );
};

export const Render = () => {
  const [data] = useState<Field[]>(mockData);
  const [submitData, setSubmitData] = useState({});

  return (
    <>
      <GeneralFormProvider>
        <Form
          fields={data}
          onSubmit={(values) => setSubmitData(values)}
          onChange={(values) => console.log(values)}
        >
          <div>
            <div
              style={{
                display: "grid",
                gridTemplateColumns: "auto auto auto auto",
                gridGap: "8px 20px",
              }}
            >
              <Form.Field id="name" label="Your name" />
              <Form.Field id="surname" label="Your surname" />
              <Form.Field
                id="email"
                label="Email"
                render={({ field }) => (
                  <input
                    {...field}
                    type="email"
                    onChange={(e) => field.onChange(e.target.value)}
                  />
                )}
              />
              <Form.Field id="active" label="Is Active?" />
            </div>
            <button style={{ marginTop: 20 }}>Submit</button>
          </div>
        </Form>
      </GeneralFormProvider>
      <br />
      {JSON.stringify(submitData)}
    </>
  );
};
