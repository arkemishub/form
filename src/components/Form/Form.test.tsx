/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { render, renderHook } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Form } from "../Form";
import { ReactNode } from "react";
import { FormConfigProvider } from "../FormConfigProvider";
import { useForm } from "../../hooks";

const fields = [
  {
    default: null,
    helper_text: null,
    id: "name",
    label: "Name",
    max_length: null,
    min_length: null,
    multiple: false,
    required: false,
    type: "string",
    values: null,
  },
  {
    default: null,
    helper_text: null,
    id: "surname",
    label: "Surname",
    max_length: null,
    min_length: null,
    multiple: false,
    required: false,
    type: "string",
    values: null,
  },
];

const fieldsWithValue = [
  {
    default: "Default value",
    helper_text: null,
    id: "name",
    label: "Name",
    max_length: null,
    min_length: null,
    multiple: false,
    required: false,
    type: "string",
    values: null,
    value: "John",
  },
  {
    default: null,
    helper_text: null,
    id: "surname",
    label: "Surname",
    max_length: null,
    min_length: null,
    multiple: false,
    required: false,
    type: "string",
    values: null,
    value: "Doe",
  },
];

export const TestFormConfigProvider = ({
  children,
}: {
  children: ReactNode;
}) => {
  return (
    <FormConfigProvider
      components={{
        string: ({ field }) => (
          <input
            {...field}
            data-testid={field.id}
            onChange={(e) => field.onChange(e.target.value)}
          />
        ),
      }}
    >
      {children}
    </FormConfigProvider>
  );
};

describe("Form", () => {
  test("should match snapshot", () => {
    const { asFragment } = render(
      <Form fields={fields}>
        <Form.Field id={"name"} />
        <Form.Field id={"surname"} />
        <button type="submit">Submit</button>
      </Form>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("should match snapshot with values", () => {
    const { asFragment } = render(
      <Form fields={fieldsWithValue}>
        <Form.Field id={"name"} />
        <Form.Field id={"surname"} />
        <button type="submit">Submit</button>
      </Form>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("should call onSubmit when submit button is clicked", async () => {
    const onSubmit = jest.fn();
    const { getByTestId } = render(
      <Form fields={fields} onSubmit={onSubmit}>
        <Form.Field id={"name"} />
        <button data-testid="form-submit" type="submit">
          Submit
        </button>
      </Form>
    );

    await userEvent.click(getByTestId("form-submit"));
    expect(onSubmit).toHaveBeenCalled();
  });

  test("should call onChange when formField value change", async () => {
    const onChange = jest.fn();
    const { getByTestId } = render(
      <TestFormConfigProvider>
        <Form fields={fields} onChange={onChange}>
          <Form.Field id={"name"} label="Name" />
        </Form>
      </TestFormConfigProvider>
    );

    await userEvent.type(getByTestId("name"), "User");
    expect(onChange).toHaveBeenCalled();
  });

  test("should render a input through the FormConfigProvider", async () => {
    const onSubmit = jest.fn();
    const { getByTestId } = render(
      <TestFormConfigProvider>
        <Form fields={fields} onSubmit={onSubmit}>
          <Form.Field id={"name"} />
          <button data-testid="form-submit" type="submit">
            Submit
          </button>
        </Form>
      </TestFormConfigProvider>
    );

    expect(getByTestId("name")).toBeInTheDocument();
  });

  test("should render a input through the components props", async () => {
    const onSubmit = jest.fn();
    const { getByTestId } = render(
      <Form
        fields={fields}
        onSubmit={onSubmit}
        components={{
          string: ({ field }) => (
            <input
              {...field}
              data-testid={field.id}
              onChange={(e) => field.onChange(e.target.value)}
            />
          ),
        }}
      >
        <Form.Field id={"name"} />
        <button data-testid="form-submit" type="submit">
          Submit
        </button>
      </Form>
    );

    expect(getByTestId("name")).toBeInTheDocument();
  });

  test("should render default component if component type doesn't exist", async () => {
    const onSubmit = jest.fn();
    const { getByText } = render(
      <Form
        fields={fields}
        onSubmit={onSubmit}
        components={{
          default: ({ field }) => <div>Component not found</div>,
        }}
      >
        <Form.Field id={"name"} />
        <button data-testid="form-submit" type="submit">
          Submit
        </button>
      </Form>
    );

    expect(getByText("Component not found")).toBeInTheDocument();
  });

  test("should Form.Field has value if Form has fields with value", async () => {
    const { getByTestId } = render(
      <TestFormConfigProvider>
        <Form fields={fieldsWithValue}>
          <Form.Field id={"name"} label="Name" />
        </Form>
      </TestFormConfigProvider>
    );

    expect(getByTestId("name")).toHaveAttribute("value", "John");
  });

  test("should get defaultValue using getFieldDefaultValue", async () => {
    const { result } = renderHook(() =>
      useForm({
        fields: fieldsWithValue.map((item) => {
          // @ts-ignore
          item.defaultValue = item.default;
          return item;
        }),
        getFieldDefaultValue: (field) => field.defaultValue,
      })
    );
    const { getByTestId } = render(
      <TestFormConfigProvider>
        <Form {...result.current.formProps}>
          <Form.Field id={"name"} label="Name" />
        </Form>
      </TestFormConfigProvider>
    );

    expect(getByTestId("name")).toHaveAttribute("value", "Default value");
  });

  test("should get methods and formProps with useHook", async () => {
    const onSubmit = jest.fn();
    const { result } = renderHook(() => useForm());
    const { register, watch, getValues, reset } = result.current.methods;
    const { getByTestId } = render(
      <TestFormConfigProvider>
        <Form fields={fields} onSubmit={onSubmit}>
          <Form.Field id={"name"} />
          {watch("name") && <Form.Field id={"surname"} />}
          <button data-testid="form-submit" type="submit">
            Submit
          </button>
        </Form>
      </TestFormConfigProvider>
    );

    expect(getByTestId("name")).toBeInTheDocument();
    expect(watch).toBeDefined();
    expect(register).toBeDefined();
    expect(reset).toBeDefined();
    expect(getValues).toBeDefined();
  });
});
