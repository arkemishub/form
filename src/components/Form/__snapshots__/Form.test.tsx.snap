// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should match snapshot 1`] = `
<DocumentFragment>
  <form>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
